bip-txhash.md

```diff
@@ -0,0 +1,269 @@
+    individual input, (at least) one byte is expected, of this byte. The
+    highest bit is used to indicate "absolute or relative" indices.
+    * If the highest bit is set to 0, it is an absolute index. The second
+      highest bit is used to indicate the amount of bytes are used to represent
+      the index.
+      * If the second-highest bit is 0, the remaining 6 bits represent the index to be selected.
+      * If the second-highest bit is 1, the remaining 6 bits, together with the 8 bits of the next
+        byte, represent the index to be selected.
+    * If the highest bit is set to 1, it is a relative index. The second highest bit is used to
+      indicate the sign of the index.
+      * If the second-highest bit is set to 0, the remaining 6 bits represent the positive relative
+        index to be selected.
+      * If the second-highest bit is set to 1, the remaining 6 bits represent the negative relative
+        index to be selected.
+
+Effectively, this allows a user to select
+* all in/outputs
+* the current input index
+* the leading in/outputs up to 8192
+* up to 64 individually selected in/outputs
+** using absolute indices up to 16384
+** using indices relative to the current input index from -64 to +64.
+
+The TxFieldSelector is invalid when
+* a byte is expected but missing
+* additional unexpected bytes are present
+* index size is set to 1 while not being necessary
+* a leading number of individual index is selected out of bounds of the in/outputs
+* individual indices are duplicated or not in increasing order
+
+These limitations are to avoid potential TxFieldSelector malleability. It is
+however allowed to use leading mode where it could be "all". This
+is important to allow for optional addition of extra inputs or outputs.
+
+//TODO(stevenroose) should we disallow individual that could be leading?
+
+
+## Resource limits
+
+* For legacy scripts and segwit, we don't add any extra resource limitations,
+  with the argumentation that `OP_CHECKTXHASHVERIFY` already requires the user
+  to provide at least 32 bytes of extra transaction size, either in the input
+  scriptSig, or the witness. Additional more complex hashes require additional
+  witness bytes. Given that `OP_CAT` is not available in this context, if a
+  malicious user tries to increase the number of TransactionHashes being
+  calculated by using opcodes like `OP_DUP`, the TxFieldSelector for all these
+  calculations is identical, so the calculation can be cached within the same
+  transaction.
+
+* For tapscript, primarily motivated by the cheaper opcode `OP_TXHASH` (it
+  doesn't require an additional 32 witness bytes be provided) and the potential
+  future addition of byte manipulation opcodes like `OP_CAT`, an additional
+  cost is specified per TransactionHash execution. Using the same validation
+  budget ("sigops budget") introduced in BIP-0342, each TransactionHash
+  decreases the validation budget by 10. If this brings the budget below zero,
```

Comment:
This needs much more justification.  Why 10?  It has an implied cost of 2 already, since you have to use the opcode and a selector.    If it has to hash a lot, hasn't someone already paid that to make such a large transaction?

Reply:
Yeah it's tricky. In practice, it has a similar amortized per-tx hash cost that sighashes have. It's hard to count those to the budget because they are amortized, it's basically hashing all the large tx fields once so that if they are repeatedly requested their hash can be used.

After the amortized hash cost, it's just a finite series of ~32-byte chunks with maximally 64 in/out which in total can have 8 fields that are each ~32 bytes. This is ~16,384 bytes max.

Then, another consideration is that it would be nice and reasonable if TXHASH+CSFS would not have a higher cost than what naturally would be placed in the witness, the 64-byte signature.

I see it like this: we have a 64-byte budget to divide over TXHASH+CSFS as I think it's reasonable that this combination doesn't cost more than 28% more than a CHECKSIG (which is 50).

So maybe it's right that TXHASH can actually cost more, something like 25 if CSFS would be priced at 35 or 40.

Reply:
This is not a signature budget, it's a hashing budget.  Perhaps we should make this a first-class citizen then?

See https://rusty.ozlabs.org/2023/12/22/script-limits-opcat.html#my-proposal-a-dynamic-limit-for-hashing